File: src/main/cpp/subsystems/superstructure/SuperstructureStateData.cpp
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

#include "subsystems/superstructure/SuperstructureStateData.h"

SuperstructureStateData::Builder::Builder()
    : pose(SuperstructurePose()),
      tunnelVolts(std::function<double()>([]() { return 0.0; })),
      gripperGoal(Dispenser::GripperGoal::IDLE),
      intakeVolts(std::function<double()>([]() { return 0.0; })),
      chariotGoal(Chariot::Goal::RETRACT),
      height(SuperstructureStateData::Height::BOTTOM), reversed(false) {}

SuperstructureStateData::Builder &
SuperstructureStateData::Builder::Pose(const SuperstructurePose &pose) {
  this->pose = pose;
  return *this;
}

SuperstructureStateData::Builder &SuperstructureStateData::Builder::TunnelVolts(
    std::function<double()> tunnelVolts) {
  this->tunnelVolts = tunnelVolts;
  return *this;
}

SuperstructureStateData::Builder &SuperstructureStateData::Builder::GripperGoal(
    Dispenser::GripperGoal gripperGoal) {
  this->gripperGoal = gripperGoal;
  return *this;
}

SuperstructureStateData::Builder &SuperstructureStateData::Builder::IntakeVolts(
    std::function<double()> intakeVolts) {
  this->intakeVolts = intakeVolts;
  return *this;
}

SuperstructureStateData::Builder &
SuperstructureStateData::Builder::ChariotGoal(Chariot::Goal chariotGoal) {
  this->chariotGoal = chariotGoal;
  return *this;
}

SuperstructureStateData::Builder &
SuperstructureStateData::Builder::Height(Height height) {
  this->height = height;
  return *this;
}

SuperstructureStateData::Builder &
SuperstructureStateData::Builder::Reversed(bool reversed) {
  this->reversed = reversed;
  return *this;
}

SuperstructureStateData SuperstructureStateData::Builder::Build() {
  SuperstructureStateData data;
  data.pose = pose;
  data.tunnelVolts = tunnelVolts;
  data.gripperGoal = gripperGoal;
  data.intakeVolts = intakeVolts;
  data.chariotGoal = chariotGoal;
  data.height = height;
  data.reversed = reversed;
  return data;
}

SuperstructureStateData::Builder SuperstructureStateData::BuilderCreate() {
  return Builder();
}

SuperstructureStateData::Builder SuperstructureStateData::ToBuilder() const {
  Builder builder;
  builder.pose = pose;
  builder.tunnelVolts = tunnelVolts;
  builder.gripperGoal = gripperGoal;
  builder.intakeVolts = intakeVolts;
  builder.chariotGoal = chariotGoal;
  builder.height = height;
  builder.reversed = reversed;
  return builder;
}

File: src/main/java/org/littletonrobotics/frc2025/subsystems/superstructure/SuperstructureStateData.java
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

package org.littletonrobotics.frc2025.subsystems.superstructure;

import edu.wpi.first.math.geometry.Rotation2d;
import java.util.function.DoubleSupplier;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.littletonrobotics.frc2025.subsystems.superstructure.dispenser.Dispenser;

@Builder(toBuilder = true, access = AccessLevel.PACKAGE)
@Getter
public class SuperstructureStateData {
  @Builder.Default
  private final SuperstructurePose pose = new SuperstructurePose(() -> 0.0, () -> Rotation2d.kZero);

  @Builder.Default private final DoubleSupplier tunnelVolts = () -> 0.0;
  @Builder.Default private final Dispenser.GripperGoal gripperGoal = Dispenser.GripperGoal.IDLE;
  @Builder.Default private final DoubleSupplier intakeVolts = () -> 0.0;
  @Builder.Default private final Height height = Height.BOTTOM;

  /** What height is the carriage above? */
  @RequiredArgsConstructor
  @Getter
  public enum Height {
    BOTTOM(0),
    FIRST_STAGE(SuperstructureConstants.stage1ToStage2Height),
    SECOND_STAGE(SuperstructureConstants.stage2ToStage3Height);

    private final double position;

    public boolean lowerThan(Height other) {
      return position <= other.position;
    }

    public boolean upperThan(Height other) {
      return position > other.position;
    }
  }
}
